src/lint/rules/algorithm-line-endings.ts

```diff
@@ -0,0 +1,203 @@
+            return;
+          }
+          if (!/,$/.test(last.contents)) {
+            report({
+              line: node.location!.start.line,
+              column: node.location!.start.column,
+              message: `expected "Else" with substeps to end with "," (found "${last.contents}")`,
+            });
+          }
+        } else {
+          if (!/(?:\.|\.\)|:)$/.test(last.contents)) {
+            report({
+              line: node.location!.start.line,
+              column: node.location!.start.column,
+              message: `expected "Else" without substeps to end with "." or ":" (found "${last.contents}")`,
+            });
+          }
+        }
+      } else if (/^Repeat/.test(initialText)) {
+        if (!hasSubsteps) {
+          report({
+            line: node.location!.start.line,
+            column: node.location!.start.column,
+            message: 'expected "Repeat" to have substeps',
+          });
+        }
+        if (node.contents.length === 1 && first.contents === 'Repeat,') {
+          return;
+        }
+        if (!/^Repeat, (?:while|until) /.test(initialText)) {
+          report({
+            line: node.location!.start.line,
+            column: node.location!.start.column,
+            message: `expected "Repeat" to start with "Repeat, while " or "Repeat, until " (found "${initialText}")`,
+          });
+        }
+        if (!/,$/.test(last.contents)) {
+          report({
+            line: node.location!.start.line,
+            column: node.location!.start.column,
+            message: 'expected "Repeat" to end with ","',
+          });
+        }
+      } else if (/^For each/.test(initialText)) {
+        if (hasSubsteps) {
+          if (!/, do$/.test(last.contents)) {
+            report({
+              line: node.location!.start.line,
+              column: node.location!.start.column,
+              message: `expected "For each" with substeps to end with ", do" (found "${last.contents}")`,
+            });
+          }
+        } else {
+          if (!/(?:\.|\.\))$/.test(last.contents)) {
+            report({
+              line: node.location!.start.line,
+              column: node.location!.start.column,
+              message: `expected "For each" without substeps to end with "." (found "${last.contents}")`,
+            });
+          }
+        }
+      } else {
+        if (hasSubsteps) {
+          if (!/:$/.test(last.contents)) {
+            report({
+              line: node.location!.start.line,
+              column: node.location!.start.column,
+              message: `expected line with substeps to end with ":" (found "${last.contents}")`,
+            });
+          }
+        } else if (!/(?:\.|\.\))$/.test(last.contents)) {
+          report({
+            line: node.location!.start.line,
+            column: node.location!.start.column,
+            message: `expected line to end with "." (found "${last.contents}")`,
```

Comment:
This message kinda implies that the only way to resolve this error is to add a `.` at the end. Any way we could say that this is conditional on the other characteristics of the line?

Reply:
Changed it to say "expected freeform line...". I don't have any better ideas.